### PR TITLE
S3 GW CopyObject to allow metadata (copy and replace)

### LINKED ISSUE #8070: [Bug]: S3 Copy doesn't allow overwrite user metadata
### What happened?

Using the S3 Gateway to copy an object doesn't copy the user metadata.
     
Steps to Reproduce:
1. Upload an object without metadata.
2. Copy the object to the same exact location, passing the metadata headers.
3. Metadata does not show up on the object.

### Expected behavior

Metadata should show up on the object.

### lakeFS version

1.31.1

### How lakeFS is installed

_No response_

### Affected clients

_No response_

### Relevant log output

_No response_

### Contact details

_No response_

### PR DESCRIPTION
**Pull request by itaiad200**
Closes #8070 

Tested by extending S3GW esti tests.

### CODE REVIEW

File: pkg/catalog/catalog.go [hunk 1]
```diff
@@ -2679,6 +2679,12 @@ func (c *Catalog) CopyEntry(ctx context.Context, srcRepository, srcRef, srcPath,
 	dstEntry.Path = destPath
 	dstEntry.AddressType = AddressTypeRelative
 	dstEntry.PhysicalAddress = c.PathProvider.NewPath()
+
+	if copyMetadata {
```
Comment (itaigilo) [inline pkg/catalog/catalog.go:2683]: I probably don't have enough context, but I find this confusing:

What's the difference between the `srcEntry.Metadata` and the `metadata` that is passed as a param?

And the bool name is `copyMetadata`, but even when it's `false`, a `metadata` is copied...

Comment (itaiad200) [reply pkg/catalog/catalog.go:2683]: Yes, I agree it's confusing. I'm changing the bool name to `copySrcMetadata`, is that better IYO?

Comment (itaiad200) [reply pkg/catalog/catalog.go]: The reason for this mysterious if/else block is that we need to support 3 behaviors:
1. Copy metadata from the source entry (metadata:`nil` ,  copySrcMetadata:`true`)
2. Use new metadata (metadata:`{...}` ,  copySrcMetadata:`false`)
3. Don't use any metadata at all to keep backward compatibility of our api (metadata:`nil` ,  copySrcMetadata:`false`)

Comment (itaigilo) [reply pkg/catalog/catalog.go]: Yes, this makes much more sense now. Thanks.

File: pkg/catalog/catalog.go [hunk 2]
```diff
@@ -2651,7 +2651,7 @@ func (c *Catalog) PrepareGCUncommitted(ctx context.Context, repositoryID string,
 }
 
 // CopyEntry copy entry information by using the block adapter to make a copy of the data to a new physical address.
-func (c *Catalog) CopyEntry(ctx context.Context, srcRepository, srcRef, srcPath, destRepository, destBranch, destPath string, opts ...graveler.SetOptionsFunc) (*DBEntry, error) {
+func (c *Catalog) CopyEntry(ctx context.Context, srcRepository, srcRef, srcPath, destRepository, destBranch, destPath string, copySrcMetadata bool, metadata Metadata, opts ...graveler.SetOptionsFunc) (*DBEntry, error) {
```
Comment (N-o-Z) [inline pkg/catalog/catalog.go:2654]: Why do we need both?
If we pass metadata - override

Comment (N-o-Z) [reply pkg/catalog/catalog.go:2654]: Also can we document it?

File: pkg/gateway/operations/operation_utils.go [hunk 3]
```diff
@@ -32,6 +32,14 @@ func amzMetaWriteHeaders(w http.ResponseWriter, metadata catalog.Metadata) {
...
 }
 
+const amzMetadataDirectiveHeaderPrefix = "X-Amz-Metadata-Directive"
+
+// Per S3 API, if the header X-Amz-Metadata-Directive is set to 'REPLACE', the metadata should be replaced.
+// Otherwise, if the value is 'COPY' or the value is missing, it should be copied.
+func shouldCopyMetadata(req *http.Request) bool {
+	return !(req.Header.Get(amzMetadataDirectiveHeaderPrefix) == "REPLACE")
+}
+
```
Comment (N-o-Z) [inline pkg/gateway/operations/operation_utils.go:42]: I think this makes it much more cleaner (and also expresses the default behavior):
```suggestion
func shouldReplaceMetadata(req *http.Request) bool {
	return req.Header.Get(amzMetadataDirectiveHeaderPrefix) == "REPLACE"
}

```

File: pkg/catalog/catalog.go (hunk 2 above)
Comment (itaiad200) [reply pkg/catalog/catalog.go:2654]: https://github.com/treeverse/lakeFS/pull/8117#discussion_r1741504142

**Review by N-o-Z — CHANGES_REQUESTED** [review]
Thanks!
Not really liking the bool + value approach, made some suggestions

File: pkg/catalog/catalog.go (hunk 2 above)
Comment (itaiad200) [reply pkg/catalog/catalog.go:2654]: In short, CopyObject from lakeFS API will pass `nil` metadata and also shouldn't copy the source metadata

Comment (N-o-Z) [reply pkg/catalog/catalog.go:2654]: You can differentiate nil and empty slice to avoid that

Comment (itaiad200) [reply pkg/catalog/catalog.go:2654]: Do you think it's better? Feel so fragile and easy to make mistakes.

Comment (N-o-Z) [reply pkg/catalog/catalog.go:2654]: Personally I feel like it's cleaner and less redundant.
I would however add simple tests to cover these 3 scenarios

Comment (N-o-Z) [reply pkg/catalog/catalog.go:2654]: Not blocking on it though

**Review by N-o-Z — APPROVED** [review]
LGTM - please note the last comment

**Pull request merged**